### PR TITLE
Multiple chord degrees (Complex and custom chord names like G7(b9,#11)/B)

### PR DESCRIPTION
**Pull request by hartman42**
This addresses chords that have multiple degrees primarily.  I brought this up in #597 and created a preview that includes the changes I have in PR #928.  At @sschmidTU suggestion, I created this PR with just the changes for chords with multiple degrees.

Chords with degrees of type "add" show up as comma-separated values in parenthesis. Alters show up in another set with "alt" in front and subtracts with "omit".

Here's the preview I put in #597 which includes the changes I have in PR #928:
image

### CODE REVIEW

**Comment by matt-uib** [comment]
@hartman42 thanks very much on this - very nice!
One thing i don't understand: how can i alter e.g. #11 in C7alt(#11, b13)?
Can't i only alter notes in the original C7 chord like 1,3,5,7 ?
Or is alter also an "add+alter" ?
But then why is there a chord like Gmaj7(#9, #11)
Long story short: could we not leave out the "alt" and just write e.g.  C7(#11, b13) ?
Another little thing: did you also add support for sus2 which is also very common? (add 2, subtract 3)
Anyway - thanks for all your help here!

**Comment by hartman42** [comment]
@matt-uib 
Yeah, the C7alt(#11,b13)... is funny...

I started with the G7alt (2nd chord, 2nd measure) and MuseScore exported the chord as a dominant with 4 degrees (b5, #5, b9, and #9).  All were of degree type "add" except the b5 which had a degree type of "alt".

So, my thought was that when I see those particular degrees with those particular types, I'll just remove them and add the word "alt" after the root chord kind.  Otherwise, it looks like G7(#5,b9,#9)(alt b5).

For the C7alt(#11,b13), "alt" is outside the parenthesis so I know it had the four degrees were removed and the chord was then just left with the #11 and b13 as "add" degree types.  It's true the #11 is redundant but sometimes I like to put it in there to emphasize the tri-tone and the b13 kind of locks me into an altered scale when improvising (half/whole - whole step scale.  1, b2, #2, 3, b5, b6, b7, 1...)

I guess the thing to note is that when "alt" appears outside the parenthesis (right after the chord kind), it's removing those four degrees.  If it appears inside, then the degrees listed with it are of that type.  For "add" degree types, there is no prefix and for the "subtract" degree type, the prefix is "omit".

The sus2 is a good one... I'll work on that now...

**Comment by sschmidTU** [comment]
Just a short note, i wouldn't trust MuseScore's MusicXML export, it omits some details and gets others wrong.
Sometimes when you export from Musescore to MusicXML and open it, there are things missing or wrong.

**Comment by hartman42** [comment]
@sschmidTU Yeah, I see that now.  It doesn't add the add/subtract degrees for dominant sus2 chords... it just calls them sus chords with the dominant degrees as opposed to a dominant chord with the add/subtract.

**Comment by no-chris** [comment]
@hartman42 You might want to consider the recent documentation update of musicxml:
https://github.com/w3c/musicxml/pull/367/files#diff-14a6e0518db71baae798a9097b49a674e2d5016d4a5cb46252fee46a52e3c29cR832

Having support for the `print-object=no` attribute by discarding those degrees would definitely make sense in the context of this PR

**Comment by sschmidTU** [comment]
A small detail: Should the function `renameCustomChordKind()` maybe be named `renameChord()` or something?
I'm not an expert on chords in MusicXML, but the definition of the `kind` element says `kind` [only] "indicates the type of chord", e.g. major. And the function not only takes the kind but also the adds, alters, subtracts, etc.
http://usermanuals.musicxml.com/MusicXML/MusicXML.htm#EL-MusicXML-kind.htm

Similarly, we could rename `addCustomChordKind()` to `addChordName()` (maybe the `custom` is unnecessary!?)
Then the class `CustomChordKind` could also be e.g. `CustomChord`.

Actually, it might be the best to integrate this CustomChordKind in the existing ChordSymbolContainer, because it basically replaces it for these custom chords, right? Then we would have two competing systems for chords.
I see now that CustomChordKind is used in ChordSymbolContainer, so maybe it can act as a subsystem there.

Looking at the CustomChordKind class, the members adds, alts, and subs aren't used anywhere in the code. Shouldn't they be? Or are they just informational?

I'll add more comments in the PR for specific code lines.

**Comment by sschmidTU** [comment]
Oh and please remove the Hancock sample (Butterfly), it's very likely copyrighted.
Also, it would be good to convert the Brooke sample from musicxml to mxl (compressed musicxml), since it's pretty large at ~150KB.

File: src/MusicalScore/VoiceData/CustomChordKind.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { ChordSymbolEnum } from "./ChordSymbolContainer";
+
+export class CustomChordKind {
+  public alternateName: string;
+  public chordKind: ChordSymbolEnum;
+  public adds: string[];
+  public alts: string[];
+  public subs: string[];
```
Comment (sschmidTU) [inline src/MusicalScore/VoiceData/CustomChordKind.ts:8]: of these class members, adds, alts and subs are currently unused. are they just informational?

File: src/MusicalScore/Graphical/EngravingRules.ts [hunk 2]
```diff
@@ -593,10 +597,57 @@ export class EngravingRules {
...
+        subs: string[],
+    ): void {
+        if (ChordSymbolEnum[chordKindText] !== undefined) {
+            this.CustomChordKinds.push(CustomChordKind.createCustomChordKind(altName, ChordSymbolEnum[chordKindText], adds, alts, subs));
+        }
+    }
+
+    public renameCustomChordKind(
+        altName: string,
+        newAltName: string,
```
Comment (sschmidTU) [inline src/MusicalScore/Graphical/EngravingRules.ts:619]: Please look at our coding conventions (how some of the other code is written) and put the function definition in one line if it has 2(+) arguments or less.

File: src/MusicalScore/Graphical/MusicSheetCalculator.ts [hunk 3]
```diff
@@ -266,7 +266,7 @@ export abstract class MusicSheetCalculator {
         // delete graphicalObjects (currently: ties) that will be recalculated, newly create GraphicalObjects streching over a single StaffEntry
         this.clearRecreatedObjects();
 
-        // this.graphicalMusicSheet.initializeActiveClefs(); // could have been changed since last render?
+        //this.graphicalMusicSheet.initializeActiveClefs(); // could have been changed since last render?
```
Comment (sschmidTU) [inline src/MusicalScore/Graphical/MusicSheetCalculator.ts:269]: unnecessary change, please undo.

File: src/MusicalScore/ScoreIO/MusicSymbolModules/ChordSymbolReader.ts [hunk 4]
```diff
@@ -138,19 +139,28 @@ export class ChordSymbolReader {
...
+                    text = ChordDegreeText[degreeType.value.trim().toLowerCase()];
+                } catch (ex) {
+                    const errorMsg: string = ITextTranslation.translateText(
+                        "ReaderErrorMessages/ChordSymbolError",
+                        "Invalid chord symbol"
+                    );
+                    musicSheet.SheetErrors.pushMeasureError(errorMsg);
+                    log.debug("InstrumentReader.readChordSymbol", errorMsg, ex);
+                    return undefined;
+                }
```
Comment (sschmidTU) [inline src/MusicalScore/ScoreIO/MusicSymbolModules/ChordSymbolReader.ts:153]: wrong indentation: reduce indentation by one block/level

File: src/MusicalScore/ScoreIO/MusicSymbolModules/ChordSymbolReader.ts [hunk 5]
```diff
@@ -138,19 +139,28 @@ export class ChordSymbolReader {
...
+            }
         }
-        return new ChordSymbolContainer(rootPitch, chordKind, bassPitch, degree, musicSheet.Rules);
+        return new ChordSymbolContainer(
+            rootPitch,
+            chordKind,
+            bassPitch,
+            degrees,
+            musicSheet.Rules
+        );
```
Comment (sschmidTU) [inline src/MusicalScore/ScoreIO/MusicSymbolModules/ChordSymbolReader.ts:164]: please put this on one line.

File: src/MusicalScore/VoiceData/ChordSymbolContainer.ts [hunk 6]
```diff
@@ -45,35 +52,94 @@ export class ChordSymbolContainer {
...
+        // I'm going to store this in a variable for now so I can evaluate it with the degrees
+        let chordKind: string = chordSymbol.getTextFromChordKindEnum(chordSymbol.ChordKind);
+        const degreeTypeAry: string[] = ["adds", "alts", "subs"];
+
+        const customChordKinds: CustomChordKind[] = chordSymbol.rules.CustomChordKinds;
+
+        for (const customKind of customChordKinds) {
+            if (
+                customKind.chordKind !== chordSymbol.chordKind
+            ) {
```
Comment (sschmidTU) [inline src/MusicalScore/VoiceData/ChordSymbolContainer.ts:104]: Please put the if with the parentheses on one line.

File: src/MusicalScore/VoiceData/ChordSymbolContainer.ts [hunk 7]
```diff
@@ -45,35 +52,94 @@ export class ChordSymbolContainer {
...
+                continue;
+            }
+
+            let check: boolean = true;
```
Comment (sschmidTU) [inline src/MusicalScore/VoiceData/ChordSymbolContainer.ts:108]: this variable shouldn't be named so generically.
Judging from the following code, something like `hasCustomKindDegrees` could be the right direction.

File: src/MusicalScore/VoiceData/ChordSymbolContainer.ts [hunk 8]
```diff
@@ -45,35 +52,94 @@ export class ChordSymbolContainer {
...
+
+            for (const degType of degreeTypeAry) {
+                for (const deg of (customKind[degType] || [])) {
+                    if (degrees[degType].indexOf(deg) < 0) {
+                        check = false;
+                        break;
                     }
-                    text += chordSymbol.ChordDegree.value.toString();
+                }
+                if (check === false) {
```
Comment (sschmidTU) [inline src/MusicalScore/VoiceData/ChordSymbolContainer.ts:117]: this can be simplified to `if (!check)` (`if (!hasCustomKindDegrees)`)

File: src/MusicalScore/VoiceData/ChordSymbolContainer.ts [hunk 9]
```diff
@@ -45,35 +52,94 @@ export class ChordSymbolContainer {
...
-                    text += chordSymbol.ChordDegree.value.toString();
-                    text += ")";
-                    break;
-                default:
+                }
+            }
+            if (check) {
+                for (const degType of degreeTypeAry) {
+                    for (const deg of (customKind[degType] || [])) {
+                        degrees[degType].splice(degrees[degType].indexOf(deg), 1);
```
Comment (sschmidTU) [inline src/MusicalScore/VoiceData/ChordSymbolContainer.ts:124]: add a comment: delete degree (ideally, also say why)

File: src/MusicalScore/VoiceData/ChordSymbolContainer.ts [hunk 10]
```diff
@@ -45,35 +52,94 @@ export class ChordSymbolContainer {
...
+                    }
+                }
+                chordKind = customKind.alternateName;
             }
         }
+
+        text += chordKind;
+
```
Comment (sschmidTU) [inline src/MusicalScore/VoiceData/ChordSymbolContainer.ts:132]: remove blank line

File: src/MusicalScore/VoiceData/ChordSymbolContainer.ts [hunk 11]
```diff
@@ -88,6 +154,7 @@ export class ChordSymbolContainer {
             text += Pitch.getNoteEnumString(transposedBassPitch.FundamentalNote);
             text += this.getTextForAccidental(transposedBassPitch.Accidental);
         }
+        console.log(text);
```
Comment (sschmidTU) [inline src/MusicalScore/VoiceData/ChordSymbolContainer.ts:157]: remove console.log.
If anything, use `log.debug("OSMD: chord text: " + text)` or similar.
(don't just log a variable, also log what this variable is, and make this log searchable in the code (by adding "chord text: ..."))

**Review by sschmidTU — CHANGES_REQUESTED** [review]
Great additions, but needs some changes, see review comments.

File: src/MusicalScore/ScoreIO/MusicSymbolModules/ChordSymbolReader.ts (hunk 4 above)
Comment (hartman42) [reply src/MusicalScore/ScoreIO/MusicSymbolModules/ChordSymbolReader.ts:153]: Actually, it looks like I had the indentation above wrong...

File: src/MusicalScore/VoiceData/ChordSymbolContainer.ts (hunk 11 above)
Comment (hartman42) [reply src/MusicalScore/VoiceData/ChordSymbolContainer.ts:157]: I need to figure out how to do that.

File: src/MusicalScore/VoiceData/CustomChordKind.ts (hunk 1 above)
Comment (hartman42) [reply src/MusicalScore/VoiceData/CustomChordKind.ts:8]: @sschmidTU - this is the only one I'm stumped on...
These are the arrays of degrees for each degree-type.  Is there another way I should do this?

File: src/MusicalScore/VoiceData/ChordSymbolContainer.ts (hunk 11 above)
Comment (sschmidTU) [reply src/MusicalScore/VoiceData/ChordSymbolContainer.ts:157]: Just write `log.debug(...)` as i wrote above, and add this to the imports at the top of the file:
`import log from "loglevel";`

File: src/MusicalScore/VoiceData/CustomChordKind.ts (hunk 1 above)
Comment (sschmidTU) [reply src/MusicalScore/VoiceData/CustomChordKind.ts:8]: Hmm, not sure on a first look, but if they are relevant in any way, shouldn't they be used somewhere? Like for `addCustomChordKind()` or the renaming of chords (give adds, alts, subs as arguments)?

Comment (hartman42) [reply src/MusicalScore/VoiceData/CustomChordKind.ts:8]: Yeah, addCustomChordKind takes five arguments... three of which are the adds, alts, and subs...

I'm also looking at your earlier comment about renaming the class... or even integrating CustomChordKind in the existing ChordSymbolContainer.  If that's your inclination, I want to give that a shot!  That might make this a moot point.

**Review by sschmidTU — APPROVED** [review]
Amazing work @hartman42 ! Merging.

This is a real game changer for chord names in OSMD, especially Jazz chords.

There are a couple minor formatting issues left, i'll quickly fix them as a follow-up after merging.

**Comment by sschmidTU** [comment]
Also the ability to rename chords (say how they should be written) in `EngravingRules.renameChord()` is fantastic for customization.

**Comment by sschmidTU** [comment]
@hartman42 one question, where does the "alt" come from here in "F#m7(alt b5)"?
image
[am7b5-sibelius_mod.xml.zip](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/files/5867729/am7b5-sibelius_mod.xml.zip)

Also, how could i rename/modify the chord text? I tried this, but nothing changed:
`this.addChordName("test", "minor-seventh", [], ["b5"], []);`

This displays without the "alt " in Sibelius, so it would be nice to be able to replicate that.

**Comment by hartman42** [comment]
> @hartman42 one question, where does the "alt" come from here in "F#m7(alt b5)"?
> image
> [am7b5-sibelius_mod.xml.zip](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/files/5867729/am7b5-sibelius_mod.xml.zip)
> 
> Also, how could i rename/modify the chord text? I tried this, but nothing changed:
> `this.addChordName("test", "minor-seventh", [], ["b5"], []);`
> 
> This displays without the "alt " in Sibelius, so it would be nice to be able to replicate that.

Try it without the dash in `minor-seventh'...

**Comment by sschmidTU** [comment]
@hartman42 Oh, i should have figured that out, thanks!

**Comment by hartman42** [comment]
> @hartman42 one question, where does the "alt" come from here in "F#m7(alt b5)"?
> image
> [am7b5-sibelius_mod.xml.zip](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/files/5867729/am7b5-sibelius_mod.xml.zip)
> 
> Also, how could i rename/modify the chord text? I tried this, but nothing changed:
> `this.addChordName("test", "minor-seventh", [], ["b5"], []);`
> 
> This displays without the "alt " in Sibelius, so it would be nice to be able to replicate that.

As for the "alt" in the parenthesis, that comes from the degree-type.  I was on the fence about making those customizable.  I've seen in osmd "omit" for subtract degree types and in other places, I've seen "no", i.e. F#7no3.  For adds, I don't add any text, for alter, I use "alt" as a prefix, and for subtract, I use "omit".

**Comment by sschmidTU** [comment]
Yeah, this is a bit of a rarer use case, we can still add that later.
Still, it would be nice to be able to customize chords so they say "no" instead of "omit".
But for now one can modify the chord names individually with the code i wrote above.

**Comment by hartman42** [comment]
My thinking exactly!  HAHA!

**Comment by hartman42** [comment]
@sschmidTU I'm glad you like it!  Thanks for the opportunity to work on the project!

**Comment by matt-uib** [comment]
Also big thanks from my side - really great work that will be much appreciated by all osmd users!
